source/extensions/transport_sockets/http_11_proxy/connect.cc

```diff
@@ -15,14 +15,15 @@ namespace Extensions {
 namespace TransportSockets {
 namespace Http11Connect {
 
-bool UpstreamHttp11ConnectSocket::isValidConnectResponse(Buffer::Instance& buffer) {
+bool UpstreamHttp11ConnectSocket::isValidConnectResponse(absl::string_view data, bool& error, unsigned long& len) {
   SelfContainedParser parser;
-  while (parser.parser().getStatus() == Http::Http1::ParserStatus::Ok &&
-         !parser.headersComplete() && buffer.length() != 0) {
-    auto slice = buffer.frontSlice();
-    int parsed = parser.parser().execute(static_cast<const char*>(slice.mem_), slice.len_);
-    buffer.drain(parsed);
+
+  len = parser.parser().execute(data.data(), data.length());
+  if (parser.parser().getStatus() == Http::Http1::ParserStatus::Error ||
+      parser.parser().statusCode() != 200) {
+    error = true;
   }
+
   return parser.parser().getStatus() != Http::Http1::ParserStatus::Error &&
          parser.headersComplete() && parser.parser().statusCode() == 200;
```

Comment:
It looks like we're repeating this same expression twice and I'm not sure I understand the difference between the return value and the `error` outparam. Could we perhaps eliminate `error`?

edit:
Oh, I see. The difference is headersComplete(). I see. Ok, how 'bout we do this:

```
bool UpstreamHttp11ConnectSocket::isValidConnectResponse(absl::string_view data, bool& headers_complete, unsigned long& header_bytes_processed) {
  SelfContainedParser parser;

  header_bytes_processed = parser.parser().execute(data.data(), data.length());
  headers_complete = parser.headersComplete();
  return parser.parser().getStatus() != Http::Http1::ParserStatus::Error &&
           parser.headersComplete() && parser.parser().statusCode() == 200;
}
```

Then below we can do something like:
```
if (!is_valid_connect_response) {
  if (headers_complete) {
    // invalid response 
  } else if (bytes_processed > x) {
    // failed to receive response quickly enought
  }
  // incomplete response
}
```

WDYT?